nbqa/handle_magics.py

```diff
@@ -210,4 +172,323 @@ class CellMagicHandler(MagicHandler):
+            {self._extract_line_magic()} {self._indent_extracted_code(extracted_code)}
+            """.strip()
+
+        return replacement_code_pattern.sub(ipython_magic, cell_source)
+
+    def _extract_code(self, replacement_code: str) -> str:
+        """Extract the python code that was present in original ipython magic.
+
+        Line magic like ``%time func(a, b)`` will get replaced in the temporary python
+        as file as
+
+        .. code:: python
+
+            if int(some_hex_token):
+                func(a, b)  # some_hex_token
+
+        This method extracts only ``func(a, b)`` from the above python code. This
+        extracted code will be written to the original notebook. This is done so that
+        changes made by formatters on those pieces of code are captured and thereby
+        preventing formatters like black reformatting the notebook every time they are
+        run on the notebook.
+
+        Parameters
+        ----------
+        replacement_code : str
+            Python code used for replacing the line magic
+
+        Returns
+        -------
+        str
+            Extracted python code
+
+        Raises
+        ------
+        AssertionError
+            Raised when unable to extract the python code
+        """
+        # from the replaced code extract the python source code
+        # that was also present in the original ipython magic
+        extract_code_pattern = MagicHandler._get_regex_pattern(
+            self._EXTRACT_CODE_REGEX_TEMPLATE, self._token
+        )
+        match_obj = extract_code_pattern.search(replacement_code)
+        if match_obj is None:  # pragma: nocover
+            raise AssertionError("Unable to extract code from the replaced python code")
+        # This is needed because if formatters like black break the code in to many
+        # lines, when the current statement exceeds the configured line length.
+        # But ipython syntax requires multiline line magics to end with a backslash
+        # By adding `\` to end of each line, line magic spanning multiple lines is
+        # parsed as single line by jupyter/ipython.
+        # For example, black formatted code like below
+        # %time np.random.randn(
+        #           100
+        #       )
+        # will become
+        # %time np.random.randn(\
+        #           100\
+        #       )
+        return re.sub("\n", "\\\n", match_obj.group(1))
+
+    def _extract_line_magic(self) -> str:
+        """Extract only the line magic of the format ``%<magic_name>``.
+
+        Returns
+        -------
+        str
+            Returns the line magic from the original source code.
+
+        Raises
+        ------
+        AssertionError
+            Raised if unable to extract the line magic from the source.
+        """
+        match_obj = re.match(r"%\w+", self._ipython_magic)
+        if match_obj is None:  # pragma: nocover
+            raise AssertionError("Unable to extract the line magic")
```

Comment:
we can do this in a separate PR, but IMO it would be nice to have tests which somehow get to these branches

Reply:
Sure. I will try to cover those in a separate PR following this one.